--- a/source/types.h
+++ b/source/types.h
@@ -36,6 +36,7 @@
 
 #ifdef GEKKO
 #include <gctypes.h>
+#include <gcbool.h>
 #include "compat.h"
 #else /* GEKKO */
 
--- a/source/ntfsfile.c
+++ b/source/ntfsfile.c
@@ -217,10 +217,10 @@
     // Unlock
     ntfsUnlock(file->vd);
 
-    return (int)fileStruct;
-}
-
-int ntfs_close_r (struct _reent *r, int fd)
+    return 0;
+}
+
+int ntfs_close_r (struct _reent *r, void *fd)
 {
     ntfs_log_trace("fd %p\n", (void *) fd);
 
@@ -253,7 +253,7 @@
     return 0;
 }
 
-ssize_t ntfs_write_r (struct _reent *r, int fd, const char *ptr, size_t len)
+ssize_t ntfs_write_r (struct _reent *r, void *fd, const char *ptr, size_t len)
 {
     ntfs_log_trace("fd %p, ptr %p, len %u\n", (void *) fd, ptr, len);
 
@@ -319,7 +319,7 @@
     return written;
 }
 
-ssize_t ntfs_read_r (struct _reent *r, int fd, char *ptr, size_t len)
+ssize_t ntfs_read_r (struct _reent *r, void *fd, char *ptr, size_t len)
 {
     ntfs_log_trace("fd %p, ptr %p, len %u\n", (void *) fd, ptr, len);
 
@@ -378,7 +378,7 @@
     return read;
 }
 
-off_t ntfs_seek_r (struct _reent *r, int fd, off_t pos, int dir)
+off_t ntfs_seek_r (struct _reent *r, void *fd, off_t pos, int dir)
 {
     ntfs_log_trace("fd %p, pos %llu, dir %i\n", (void *) fd, pos, dir);
 
@@ -406,7 +406,7 @@
 
     return position;
 }
-int ntfs_fstat_r (struct _reent *r, int fd, struct stat *st)
+int ntfs_fstat_r (struct _reent *r, void *fd, struct stat *st)
 {
     ntfs_log_trace("fd %p\n", (void *) fd);
 
@@ -431,7 +431,7 @@
     return ret;
 }
 
-int ntfs_ftruncate_r (struct _reent *r, int fd, off_t len)
+int ntfs_ftruncate_r (struct _reent *r, void *fd, off_t len)
 {
     ntfs_log_trace("fd %p, len %llu\n", (void *) fd, (u64) len);
 
@@ -498,7 +498,7 @@
     return 0;
 }
 
-int ntfs_fsync_r (struct _reent *r, int fd)
+int ntfs_fsync_r (struct _reent *r, void *fd)
 {
     ntfs_log_trace("fd %p\n", (void *) fd);
 
--- a/source/ntfsfile.h
+++ b/source/ntfsfile.h
@@ -53,13 +53,13 @@
 
 /* Gekko devoptab file routines for NTFS-based devices */
 extern int ntfs_open_r (struct _reent *r, void *fileStruct, const char *path, int flags, int mode);
-extern int ntfs_close_r (struct _reent *r, int fd);
-extern ssize_t ntfs_write_r (struct _reent *r, int fd, const char *ptr, size_t len);
-extern ssize_t ntfs_read_r (struct _reent *r, int fd, char *ptr, size_t len);
-extern off_t ntfs_seek_r (struct _reent *r, int fd, off_t pos, int dir);
-extern int ntfs_fstat_r (struct _reent *r, int fd, struct stat *st);
-extern int ntfs_ftruncate_r (struct _reent *r, int fd, off_t len);
-extern int ntfs_fsync_r (struct _reent *r, int fd);
+extern int ntfs_close_r (struct _reent *r, void *fd);
+extern ssize_t ntfs_write_r (struct _reent *r, void *fd, const char *ptr, size_t len);
+extern ssize_t ntfs_read_r (struct _reent *r, void *fd, char *ptr, size_t len);
+extern off_t ntfs_seek_r (struct _reent *r, void *fd, off_t pos, int dir);
+extern int ntfs_fstat_r (struct _reent *r, void *fd, struct stat *st);
+extern int ntfs_ftruncate_r (struct _reent *r, void *fd, off_t len);
+extern int ntfs_fsync_r (struct _reent *r, void *fd);
 
 #endif /* _NTFSFILE_H */
 
--- a/source/ntfsinternal.c
+++ b/source/ntfsinternal.c
@@ -292,7 +292,7 @@
 
     // If the entry was found and it has reparse data then parse its true path;
     // this resolves the true location of symbolic links and directory junctions
-    if (ni && (ni->flags & FILE_ATTR_REPARSE_POINT)) {
+    if (reparseLevel>=0 && ni && (ni->flags & FILE_ATTR_REPARSE_POINT)) {
         if (ntfs_possible_symlink(ni)) {
 
             // Sanity check, give up if we are parsing to deep
--- a/source/ntfsdir.c
+++ b/source/ntfsdir.c
@@ -634,3 +634,54 @@
 
     return 0;
 }
+
+
+int ntfs_lstat_r (struct _reent *r, const char *path, struct stat *st)
+{
+    // Short circuit cases were we don't actually have to do anything
+    if (!st || !path)
+        return 0;
+
+    ntfs_log_trace("path %s, st %p\n", path, st);
+
+    ntfs_vd *vd = NULL;
+    ntfs_inode *ni = NULL;
+
+    // Get the volume descriptor for this path
+    vd = ntfsGetVolume(path);
+    if (!vd) {
+        r->_errno = ENODEV;
+        return -1;
+    }
+
+    if(strcmp(path, ".") == 0 || strcmp(path, "..") == 0)
+    {
+        memset(st, 0, sizeof(struct stat));
+        st->st_mode = S_IFDIR;
+        return 0;
+    }
+
+    // Lock
+    ntfsLock(vd);
+
+    // Find the entry
+    ni = ntfsParseEntry(vd, path, -1);
+    if (!ni) {
+        r->_errno = errno;
+        ntfsUnlock(vd);
+        return -1;
+    }
+
+    // Get the entry stats
+    int ret = ntfsStat(vd, ni, st);
+    if (!ret && (ni->flags & FILE_ATTR_REPARSE_POINT)) st->st_mode = S_IFLNK;
+    if (ret)
+        r->_errno = errno;
+
+    // Close the entry
+    ntfsCloseEntry(vd, ni);
+
+    ntfsUnlock(vd);
+
+    return ret;
+}
--- a/source/ntfsdir.h
+++ b/source/ntfsdir.h
@@ -50,6 +50,7 @@
 void ntfsCloseDir (ntfs_dir_state *file);
 
 /* Gekko devoptab directory routines for NTFS-based devices */
+extern int ntfs_lstat_r (struct _reent *r, const char *path, struct stat *st);
 extern int ntfs_stat_r (struct _reent *r, const char *path, struct stat *st);
 extern int ntfs_link_r (struct _reent *r, const char *existing, const char *newLink);
 extern int ntfs_unlink_r (struct _reent *r, const char *name);
--- a/source/ntfs.c
+++ b/source/ntfs.c
@@ -64,7 +64,8 @@
     ntfs_statvfs_r,
     ntfs_ftruncate_r,
     ntfs_fsync_r,
-    NULL /* Device data */
+    NULL /* Device data */,
+    NULL, NULL, NULL, ntfs_lstat_r
 };
 
 void ntfsInit (void)
