--- a/source/unzip.c
+++ b/source/unzip.c
@@ -147,7 +147,7 @@
     int encrypted;
 #    ifndef NOUNCRYPT
     unsigned long keys[3];     /* keys defining the pseudo-random sequence */
-    const unsigned long* pcrc_32_tab;
+    const z_crc_t* pcrc_32_tab;
 #    endif
 } unz_s;
 
--- a/source/zip.c
+++ b/source/zip.c
@@ -131,7 +131,7 @@
     int  encrypt;
 #ifndef NOCRYPT
     unsigned long keys[3];     /* keys defining the pseudo-random sequence */
-    const unsigned long* pcrc_32_tab;
+    const z_crc_t* pcrc_32_tab;
     int crypt_header_size;
 #endif
 } curfile_info;
--- a/source/crypt.h
+++ b/source/crypt.h
@@ -32,7 +32,7 @@
 /***********************************************************************
  * Return the next byte in the pseudo-random sequence
  */
-static int decrypt_byte(unsigned long* pkeys, const unsigned long* pcrc_32_tab)
+static int decrypt_byte(unsigned long* pkeys, const z_crc_t* pcrc_32_tab)
 {
     (void) pcrc_32_tab; /* avoid "unused parameter" warning */
     unsigned temp;  /* POTENTIAL BUG:  temp*(temp^1) may overflow in an
@@ -46,7 +46,7 @@
 /***********************************************************************
  * Update the encryption keys with the next byte of plain text
  */
-static int update_keys(unsigned long* pkeys,const unsigned long* pcrc_32_tab,int c)
+static int update_keys(unsigned long* pkeys,const z_crc_t* pcrc_32_tab,int c)
 {
     (*(pkeys+0)) = CRC32((*(pkeys+0)), c);
     (*(pkeys+1)) += (*(pkeys+0)) & 0xff;
@@ -63,7 +63,7 @@
  * Initialize the encryption keys and the random header according to
  * the given password.
  */
-static void init_keys(const char* passwd,unsigned long* pkeys,const unsigned long* pcrc_32_tab)
+static void init_keys(const char* passwd,unsigned long* pkeys,const z_crc_t* pcrc_32_tab)
 {
     *(pkeys+0) = 305419896L;
     *(pkeys+1) = 591751049L;
@@ -93,7 +93,7 @@
     unsigned char *buf;         /* where to write header */
     int bufSize;
     unsigned long* pkeys;
-    const unsigned long* pcrc_32_tab;
+    const z_crc_t* pcrc_32_tab;
     unsigned long crcForCrypting;
 {
     int n;                       /* index in random header */
--- a/source/ioapi.c
+++ b/source/ioapi.c
@@ -145,8 +145,7 @@
         break;
     default: return -1;
     }
-    ret = 0;
-    fseek((FILE *)stream, offset, fseek_origin);
+    ret = fseek((FILE *)stream, offset, fseek_origin);
     return ret;
 }
 
--- a/source/unzip.c
+++ b/source/unzip.c
@@ -428,15 +428,13 @@
         return NULL;
 
     central_pos = unzlocal_SearchCentralDir(&us.z_filefunc,us.filestream);
-    if (central_pos==0)
-        err=UNZ_ERRNO;
 
     if (ZSEEK(us.z_filefunc, us.filestream,
                                       central_pos,ZLIB_FILEFUNC_SEEK_SET)!=0)
         err=UNZ_ERRNO;
 
     /* the signature, already checked */
-    if (unzlocal_getLong(&us.z_filefunc, us.filestream,&uL)!=UNZ_OK)
+    if (unzlocal_getLong(&us.z_filefunc, us.filestream,&uL)!=UNZ_OK || uL!=0x06054b50)
         err=UNZ_ERRNO;
 
     /* number of this disk */
--- a/source/zip.c
+++ b/source/zip.c
@@ -561,15 +561,13 @@
         uLong size_comment;
 
         central_pos = ziplocal_SearchCentralDir(&ziinit.z_filefunc,ziinit.filestream);
-        if (central_pos==0)
-            err=ZIP_ERRNO;
 
         if (ZSEEK(ziinit.z_filefunc, ziinit.filestream,
                                         central_pos,ZLIB_FILEFUNC_SEEK_SET)!=0)
             err=ZIP_ERRNO;
 
         /* the signature, already checked */
-        if (ziplocal_getLong(&ziinit.z_filefunc, ziinit.filestream,&uL)!=ZIP_OK)
+        if (ziplocal_getLong(&ziinit.z_filefunc, ziinit.filestream,&uL)!=ZIP_OK || uL!=0x06054b50)
             err=ZIP_ERRNO;
 
         /* number of this disk */
